pathfinderlib/main.py

```diff
@@ -25,33 +63,212 @@ def make_fastq_to_fasta_command(self, filename):
             ]
 
     def fastq_to_fasta(self, filename):
-        fasta = tempfile.NamedTemporaryFile(delete=False)
-        command = self.make_fastq_to_fasta_command(filename)
+        fasta = tempfile.NamedTemporaryFile()
+        command = self._make_fastq_to_fasta_command(filename)
         subprocess.check_call(command, stdout=fasta, stderr=subprocess.STDOUT)
-        return fasta.name
+        return fasta
 
-    def make_blastx_command(self, R, rex, output):
-        out_dir = os.path.join(self.config["humann_fp"], "input", output + rex)
-        return [
-            "blastx", "-outfmt", "6",
-            "-db", self.config["kegg_fp"],
-            "-query", R, "-out", out_dir
-            ]
+    def make_command(self, R, output):
+        raise NotImplementedError("Create and override method in child tool")
 
-    def run_blastx(self, R, rex, output):
+    def run(self, R, out_dir):
         r_fasta = self.fastq_to_fasta(R)
-        command = self.make_blastx_command(r_fasta, rex, out_dir)
+        output_fp = self.make_output_fp(out_dir, R)
+        command = self.make_command(r_fasta.name, output_fp)
         subprocess.check_call(command, stderr=subprocess.STDOUT)
-        os.remove(r_fasta)
+        return output_fp+'.m8' if self.search_method.lower()=='rapsearch' else output_fp
+
+class Blast(_Alignment):
+   def __init__(self, search_method, kegg_fp, num_threads):
+       super(Blast, self).__init__(search_method, kegg_fp, num_threads)
+
+   def make_output_fp(self, out_dir, R):
+       return os.path.join(out_dir, os.path.basename(os.path.splitext(R)[0]+'.blast'))
+            
+   def make_command(self, R, output_fp):
+       return [
+           "blastx", "-outfmt", "6", "-num_threads", str(self.num_threads),
+           "-db", self.kegg_fp,
+           "-query", R, "-out", output_fp
+           ]
+
+   def make_index(self):
+       raise IOError("Protein fasta file  can't be found. Please check kegg_fp in the config file")
+
+   def index_exists(self):
+       return os.path.exists(self.kegg_fp)
+
+class RapSearch(_Alignment):
+    def __init__(self, search_method, kegg_fp, num_threads, kegg_idx_fp, rap_search_fp):
+        super(RapSearch, self).__init__(search_method, kegg_fp, num_threads)
+        self.kegg_idx_fp = kegg_idx_fp
+        self.rap_search_fp = rap_search_fp
+
+    def make_output_fp(self, out_dir, R):
+        return os.path.join(out_dir, os.path.basename(os.path.splitext(R)[0]))
+                
+    def make_command(self, R, output_fp):
+        return [
+            self.rap_search_fp, "-q", R,
+            "-d", self.kegg_idx_fp,
+            "-o", output_fp,
+            "-z", str(self.num_threads), "-s", "f"
+            ]
+
+    def make_index(self):
+        cmd = [
+            os.path.join(os.path.dirname(self.rap_search_fp),'prerapsearch'),
+            "-d", self.kegg_fp,
+            "-n", self.kegg_idx_fp
+            ]
+        subprocess.check_call(cmd, stderr=subprocess.STDOUT)
```

Comment:
OK to leave this in for now, but eventually we'd like to build the indexes elsewhere and assume they are present when this tool is run.